quasardb/reader/ts_row.hpp

```diff
@@ -0,0 +1,247 @@
+    ts_row() noexcept
+        : _local_table{nullptr}
+    {}
+
+    ts_row(qdb_local_table_t local_table) noexcept
+        : _local_table{local_table}
+    {}
+
+    bool operator==(const ts_row & rhs) const noexcept
+    {
+        auto tie = [](const auto & ts) { return std::tie(ts.tv_sec, ts.tv_nsec); };
+
+        // Since our row doesn't hold any intrinsic data itself and is merely
+        // an indirection to the data in the local table, it doesn't make a lot
+        // of sense to compare it with another other than comparing the timestamps
+        // and the local table references.
+        return (tie(_timestamp) == tie(rhs._timestamp)) && (_local_table == rhs._local_table);
+    }
+
+    py::handle timestamp() const noexcept
+    {
+        return numpy::to_datetime64(_timestamp);
+    }
+
+    /**
+     * Not exposed through Python, but allows the read_row operation to write the underlying
+     * timestamp object directly.
+     */
+    constexpr qdb_timespec_t & mutable_timestamp() noexcept
+    {
+        return _timestamp;
+    }
+
+protected:
+    qdb_local_table_t _local_table;
+    qdb_timespec_t _timestamp;
+};
+
+/**
+ * Our 'fast' row is a pure, lazy list-type row that uses a column's offset for
+ * constant-time access to the column values. E.g. row[0], row[1] which will return
+ * a (short-lived) ts_value.
+ */
+class ts_fast_row : public ts_row
+{
+public:
+    ts_fast_row() noexcept(std::is_default_constructible<ts_row>::value) = default;
+
+    ts_fast_row(qdb_local_table_t local_table, const ts_columns_t & columns) noexcept
+        : ts_row(local_table)
+        , _columns(columns)
+    {}
+
+    std::vector<py::object> copy() const
+    {
+        std::vector<py::object> res;
+        res.reserve(_columns.size());
+
+        for (auto index = 0; index < _columns.size(); ++index)
+        {
+            // By first casting the value, we create a copy of the concrete type
+            // (e.g. the int64 or the blob) rather than the reference into the local
+            // table.
+            res.push_back(py::cast(get_item(index), py::return_value_policy::move));
+        }
+
+        return res;
+    }
+
+    ts_value get_item(int64_t index) const
+    {
+        // TODO(leon): we construct this object every time; should be without any heap
+        // allocations,but can we do without this?
+        return ts_value(_local_table, index, _columns.at(index).type);
+    }
+
+    void set_item(int64_t /* index */, int64_t /* value */)
+    {
+        // not implemented
+    }
+
+    std::string repr() const
+    {
+        // This is inefficient because we first cast everything to Python objects, get the
+        // string representation and then cast back to c++ strings, but it's correct and
+        // effective, and it's used for debugging purposes only.
+        auto xs         = copy();
+        std::string kvs = std::accumulate(xs.cbegin(), xs.cend(), std::string(), [](std::string & acc, auto x) {
```

Comment:
```suggestion
        std::string kvs = std::accumulate(xs.cbegin(), xs.cend(), std::string(), [](std::string && acc, auto x) {
```
Use rvalue and move below.